File: src/org/likexin/binarytree/PostorderTraversal.java
package org.likexin.binarytree;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * http://www.lintcode.com/en/problem/binary-tree-postorder-traversal/
 *
 * @author Shingo Lee
 */
public class PostorderTraversal {

    /**
     * 后序遍历：左右根
     *
     * @param root 二叉树根结点
     * @return 后序遍历集合
     */
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode prev = null;
        TreeNode curr = root;

        if (root == null) {
            return result;
        }

        stack.push(root);
        while (!stack.isEmpty()) {
            curr = stack.peek();
            if (prev == null || prev.left == curr || prev.right == curr) {
                if (curr.left != null) {
                    stack.push(curr.left);
                } else if (curr.right != null) {
                    stack.push(curr.right);
                }
            } else if (curr.left == prev) {
                if (curr.right != null) {
                    stack.push(curr.right);
                }
            } else {
                result.add(curr.val);
                stack.pop();
            }
            prev = curr;
        }
        return result;
    }
}

File: Python/binarytree/postorderTraversal.py
# -*- coding:UTF-8 -*-
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # solution 1: recursion
        if not root:
            return []
        ans = [root.val]
        return self.postorderTraversal(root.left) + self.postorderTraversal(root.right) + ans
        
        
        # solution 2: loop
        if not root:
            return []
        ans, stack = [], []
        cur, pre = None, None
        stack.append(root)
        while stack:
            cur = stack.pop()
            if (cur.left == None and cur.right == None) or (pre and (pre == cur.left or pre == cur.right)):
                ans.append(cur.val)
                pre = cur
            else:
                stack.append(cur)
                if cur.right:
                    stack.append(cur.right)
                if cur.left:
                    stack.append(cur.left)
        return ans
        
        